sherpa/optmethods/optscipy.py

```diff
@@ -0,0 +1,502 @@
+`the scipy docs for global optimizers <https://docs.scipy.org/doc/scipy/tutorial/optimize.html#global-optimization>`_
+for details.
+"""
+from collections.abc import Callable
+import functools
+import inspect
+from typing import Any
+
+
+import numpy as np
+from sherpa.optmethods.buildin import OptMethod
+from sherpa.utils.types import (ArrayType,
+                                OptFunc, OptReturn,
+                                StatFunc)
+from sherpa.models.parameter import hugeval
+
+# Will be filled out programmatically at the bottom of this file
+__all__ = []
+
+
+def convert_bounds_to_scipy(parmins: ArrayType,
+                            parmaxes: ArrayType,
+                            requires_finite_bounds: bool) -> list[tuple[float | None, float | None]] | None:
+    """Convert parmins and parmaxes to format used by `scipy.optimize` for bounds.
+
+    Parameters
+    ----------
+    parmins : ArrayType
+        The minimum values for the parameters.
+    parmaxes : ArrayType
+        The maximum values for the parameters.
+    requires_finite_bounds : bool
+        If True, the scipy function requires finite bounds.
+
+    Returns
+    -------
+    list[tuple[float | None, float | None]] | None
+        The bounds in the format used by scipy.optimize, or None if
+        there are no bounds.
+    """
+    if np.allclose(parmins, -hugeval) and np.allclose(parmaxes, hugeval):
+        bounds = None
+    else:
+        bounds = [(None if pmin == -hugeval else pmin,
+                 None if pmax == hugeval else pmax)
+                for pmin, pmax in zip(parmins, parmaxes)]
+    if requires_finite_bounds:
+        if None in np.array(bounds):
+            raise ValueError("The scipy function requires finite bounds, but "
+                             "the Sherpa model has some bounds set to HUGEVAL.")
+    return bounds
+
+
+def wrap_scipy_fcn(func: Callable,
+                   requires_finite_bounds: bool) -> OptFunc:
+    """Wrap a function in scipy.optimize to the Sherpa interface.
+    """
+    sig = inspect.signature(func)
+
+    @functools.wraps(func)
+    def fcn(stat: StatFunc,
+         x0: np.ndarray,
+         xmin: np.ndarray,
+         xmax: np.ndarray,
+         **kwargs) -> OptReturn:
+
+        def stat_wrapper(x):
+            # The function is called with the parameters
+            # and returns the statistic and per-bin values
+            return stat(x)[0]
+
+        converted_args: dict = {}
+        if 'x0' in sig.parameters.keys():
+            converted_args['x0'] = x0
+        if 'bounds' in sig.parameters.keys():
+            converted_args['bounds'] = convert_bounds_to_scipy(xmin, xmax,
+                                                                requires_finite_bounds)
+
+        result = func(stat_wrapper, **converted_args, **kwargs)
+        for arg in ['bounds', 'ranges']:
+            if arg in converted_args:
+                result[f'input_{arg}'] = converted_args[arg]
```

Comment:
From my reading of this code, `converted_args` cann only contain `x0` and `bounds` at most, so why are we also checking tor `ranges` here?